source/postprocess/sealevel.cc

```diff
@@ -0,0 +1,655 @@
+      const double topography = this->get_geometry_model().height_above_reference_surface(position);
+      
+      Point<dim-1> internal_position;
+      const std::array<double,dim> spherical_position = aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(position);
+      for (unsigned int d = 1; d < dim; ++d)
+        internal_position[d-1] = spherical_position[d];
+
+      const double topography_init = topography_lookup->get_data(internal_position,0);
+      int ocean_mask = 0;
+      if (topography_init > 0.0)
+        ocean_mask = 1;
+      else
+        ocean_mask = 0;
+      
+      const double nonuniform_sealevel_change = (geoid_displacement-topography)*ocean_mask;
+
+      return nonuniform_sealevel_change;
+    }
+
+    template <int dim>
+    double
+    SeaLevel<dim>::compute_sealevel_offset(const double &/*outer_radius*/) const
+    {
+      Assert(false, ExcNotImplemented());
+    }
+
+    template <>
+    double
+    SeaLevel<3>::compute_sealevel_offset(const double &outer_radius) const
+    {
+      const int dim = 3;
+
+      const unsigned int quadrature_degree = this->introspection().polynomial_degree.temperature;
+      const QGauss<dim-1> quadrature_formula_face(quadrature_degree);
+
+      FEFaceValues<dim> fe_face_values (this->get_mapping(),
+                                        this->get_fe(),
+                                        quadrature_formula_face,
+                                        update_values |
+                                        update_quadrature_points |
+                                        update_JxW_values);
+
+
+      // Vectors to store the location, infinitesimal area, and topography/geoid displacement/ocean mask/ice height associated with each quadrature point of each surface cell.
+      std::vector<std::pair<Point<dim>,std::pair<double,double>>> topo_stored_values;
+      std::vector<std::pair<Point<dim>,std::pair<double,double>>> geoid_stored_values;
+      std::vector<std::pair<Point<dim>,std::pair<double,double>>> oceanmask_stored_values;
+      std::vector<std::pair<Point<dim>,std::pair<double,double>>> iceheight_stored_values;
+
+      double integral_oceanmask = 0;
+      double integral_iceheight = 0;
+      double integral_topo_geoid = 0;
+
+      // Loop over all of the boundary cells and if one is at the
+      // surface, evaluate the topography/geoid displacement/ocean mask/ice height there.
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        if (cell->is_locally_owned() && cell->at_boundary())
+          {
+            unsigned int face_idx = numbers::invalid_unsigned_int;
+            bool at_upper_surface = false;
+            {
+              for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+                {
+                  if (cell->at_boundary(f) && cell->face(f)->boundary_id() == this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"))
```

Comment:
Please make the same change here that I did for the geoid in #4223. It does not change the function, but it makes it faster.